File: video-club-rental/typescript/tests/RecordingNotifier.ts
import type { Notifier } from '../src/Notifier.js';
import type { User } from '../src/User.js';

export interface Notification { user: User; message: string; }

export class RecordingNotifier implements Notifier {
  private _sent: Notification[] = [];
  get sent(): readonly Notification[] { return this._sent; }
  send(user: User, message: string): void { this._sent.push({ user, message }); }
  notificationsFor(user: User): Notification[] {
    return this._sent.filter((n) => n.user === user);
  }
}

File: video-club-rental/python/tests/recording_notifier.py
from __future__ import annotations

from dataclasses import dataclass

from video_club_rental import User


@dataclass
class Notification:
    user: User
    message: str


class RecordingNotifier:
    def __init__(self) -> None:
        self._sent: list[Notification] = []

    @property
    def sent(self) -> list[Notification]:
        return list(self._sent)

    def send(self, user: User, message: str) -> None:
        self._sent.append(Notification(user, message))

    def notifications_for(self, user: User) -> list[Notification]:
        return [n for n in self._sent if n.user is user]
